sphinx/content/ch1.md

```diff
@@ -0,0 +1,1249 @@
+If we store MRI values over time, we might need a 4D NumPy array.
+
+For now, we'll stick to 2D data.
+Later chapters will introduce higher-dimensional data and will teach you to write code that works for data of any number of dimensions.
+
+### Why Use ndarrays Instead of Python Lists?
+
+Arrays are fast because they enable *vectorized* operations, written in the low-level language C, that act on the whole array.
+Say you have a list and you want to multiply every element in the list by five.
+A standard Python approach would be to write a loop that iterates over the
+elements of the list and multiply each one by five.
+However, if your data were instead represented as an array,
+you can multiply every element in the array by five in a single bound.
+Behind the scenes, the highly-optimized NumPy library is doing the iteration as fast as possible.
+
+```python
+import numpy as np
+
+# Create an ndarray of integers in the range
+# 0 up to (but not including) 1,000,000
+array = np.arange(1e6)
+
+# Convert it to a list
+list_array = array.tolist()
+```
+
+Let's compare how long it takes to multiply all the values in the array by five,
+using the IPython `timeit` magic function. First, when the data is in a list:
+
+```ipython
+%timeit -n10 y = [val * 5 for val in list_array]
+```
+
+Now, using NumPy's built-in *vectorized* operations:
+
+```ipython
+%timeit -n10 x = array * 5
+```
+
+Over 50 times faster, and more concise, too!
+
+Arrays are also size efficient.
+In Python, each element in a list is an object and is given a healthy memory allocation (or is that unhealthy?).
+In contrast, in arrays, each element takes up just the necessary amount of memory.
+For example, an array of 64-bit integers takes up exactly 64-bits per element, plus some very small overhead for array metadata, such as the `shape` attribute we discussed above.
+This is generally much less than would be given to objects in a Python list.
+```{note}
+If you're interested in digging into how Python memory allocation works,
+check out Jake VanderPlas's blog post:
+["Why Python Is Slow: Looking Under the Hood"](https://jakevdp.github.io/blog/2014/05/09/why-python-is-slow/).
+```
+
+Plus, when computing with arrays, you can also use *slices* that subset the array *without copying the underlying data*.
+
+```python
+# Create an ndarray x
+x = np.array([1, 2, 3], np.int32)
+print(x)
+```
+
+```python
+# Create a "slice" of x
+y = x[:2]
+print(y)
+```
+
+```python
+# Set the first element of y to be 6
+y[0] = 6
+print(y)
+```
+
+Notice that although we edited `y`, `x` has also changed, because `y` was referencing the same data!
+
+```python
+# Now the first element in x has changed to 6!
+print(x)
+```
+
+This means you have to be careful with array references.
+If you want to manipulate the data without touching the original, it's easy to make a copy:
+
+```python
+y = np.copy(x[:2])
+```
+
+(sec:vectorization)=
```

Comment:
This syntax is pretty odd.  Why not:

{section=vectorization} or ``` {section}`vectorization` ``` ?

Reply:
I don't think the syntax is necessarily finalized - there [was some discussion](https://github.com/ExecutableBookProject/MyST-Parser/issues/10) though I'm not sure what the current status is.